### PR TITLE
add generic V&V reporting

### PR DESCRIPTION
**Pull request by Rikkola**
Main ticket:
https://issues.redhat.com/browse/DROOLS-5203

Changes to generic Validator:
* Renaming from DRL specific objects and method to generic names
* Modified Redundancy/Subsumption check to also cover Overlaps
* Added the concepts of Intervals and Bounds

This adds in as experimental:
* Overlaps
* Misleading rows
* Masked rows
* Subsumptant rows
* Range gap checks

https://github.com/kiegroup/drools/pull/2828
https://github.com/kiegroup/kie-wb-common/pull/3250
https://github.com/kiegroup/drools-wb/pull/1342

### CODE REVIEW

**Comment by Rikkola** [comment]
jenkins please restest this

**Comment by Rikkola** [comment]
The analysis report still has the old overlaps in it, I'm planning to remove them, but they are atm there to prove I have the same coverage.

File: kie-dmn/kie-dmn-validation/src/main/java/org/kie/dmn/validation/dtanalysis/DMNDTAnalyser.java [hunk 1]
```diff
@@ -141,6 +141,13 @@ private DTAnalysis dmnDTAnalysis(DMNModel model, DecisionTable dt) {
         analysis.compute2ndNFViolations();
         LOG.debug("computeHitPolicyRecommender");
         analysis.computeHitPolicyRecommender();
+
```
Comment (Rikkola) [inline kie-dmn/kie-dmn-validation/src/main/java/org/kie/dmn/validation/dtanalysis/DMNDTAnalyser.java:158]: This bit below will replace the lines 124-131 above. The original overlaps, misleading and masked rows are still resolved, but not reported or used in the reports added byt `addIssues` below.

**Comment by tarilabs** [comment]
not sure about my involvement as this does not fill the agnostic API for DMN gaps and overlaps which enables any other checks. If I missed something don't hesitate to let me know

File: kie-dmn/kie-dmn-validation/src/test/java/org/kie/dmn/validation/dtanalysis/StringWithoutEnumNoGapTest.java [hunk 2]
```diff
@@ -93,21 +94,21 @@ public void test() {
...
-
-        // MisleadingRules count.
-        assertThat(analysis.getMisleadingRules(), hasSize(2));
-        List<MisleadingRule> misleadingRules = Arrays.asList(new MisleadingRule(3, 1),
-                                                             new MisleadingRule(3, 2));
-        assertThat(misleadingRules, hasSize(2));
-        assertThat(analysis.getMisleadingRules(), contains(misleadingRules.toArray()));
-        assertTrue("It should contain DMNMessage for the MisleadingRule",
-                   validate.stream().anyMatch(p -> p.getMessageType().equals(DMNMessageType.DECISION_TABLE_MISLEADING_RULE)));
+//        assertThat(analysis.getMaskedRules(), hasSize(2));
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-validation/src/test/java/org/kie/dmn/validation/dtanalysis/StringWithoutEnumNoGapTest.java]: ???

File: kie-dmn/kie-dmn-validation/src/test/java/org/kie/dmn/validation/dtanalysis/SameMsgInAllAPITest.java [hunk 3]
```diff
@@ -114,12 +115,12 @@ private void verify(List<DMNMessage> validate) {
...
 
         // MaskedRules count.
-        assertThat(analysis.getMaskedRules(), hasSize(2));
-        List<MaskedRule> maskedRules = Arrays.asList(new MaskedRule(2, 1),
-                                                     new MaskedRule(4, 3));
-        assertThat(maskedRules, hasSize(2));
-        assertThat(analysis.getMaskedRules(), contains(maskedRules.toArray()));
-        assertTrue("It should contain DMNMessage for the MaskedRule",
-                   validate.stream().anyMatch(p -> p.getMessageType().equals(DMNMessageType.DECISION_TABLE_MASKED_RULE)));
+//        assertThat(analysis.getMaskedRules(), hasSize(2));
```
Comment (tarilabs) [inline kie-dmn/kie-dmn-validation/src/test/java/org/kie/dmn/validation/dtanalysis/SameMsgInAllAPITest.java]: ??

**Comment by Rikkola** [comment]
@tarilabs I think your find gaps implementation will stay in. However I also think both mine and your style of finding the gaps could be done better in a 3rd way. 

This PR starts to remove your overlaps codes. This replaces your misleading, masked and overlapping row reporting. Overlap data will still be there since your redundancy checks use it, but both overlaps and the redundancy are on the list that I'm aiming to replace.

I marked you as a reviewer since I think you might find this interesting, but @manstis is already taking a look at the implementation. So if you are not against this then we are good.

**Comment by tarilabs** [comment]
Q1. some test have commented out results. If you merge this code, how is this supposed to be checked?

Q2. so are you stepping in as the maintainer for `kie-dmn-validation`?

Q3. How the new impl which you are putting forward will support the future DMN checks which are missing even in my impl. at the moment, please?

**Comment by Rikkola** [comment]
Answer 1. 
There is still work to be done on the tests. I'll get the commented tests updated for you to prove they work, but like I mentioned in PR description old checks and and some unit tests coverage is still in. This is to prove the results match.

Answer 2.
Can answer who maintains this. I can do it, but the work here is to get in what Mark has been asking me to deliver. 

Answer 3.
To answer this I could use more information on if you think these changes block future features that possibly depended on the redundancy/overlaps that you resolved? For example I'm not planning to replace all your codes, just replacing the ones that are clearly redundant. If the code you have works at the moment, I can't see a reason for me to spend yet more time replacing it. My goal is to remove feature redundancy and hopefully in the long run get some good wins for both DRL and DMN validation with this merge. Any future development can be based on my generic validator or added as a separate unit.

**Comment by Rikkola** [comment]
Checking impact of the changes in a PR. I expect I need to still fix few things before moving out of Draft-status.

**Pull request closed without merging**